pkg/config/interface.go

```diff
@@ -0,0 +1,112 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"sigs.k8s.io/kubebuilder/v3/pkg/model/resource"
+)
+
+// Config defines the interface that project configuration types must follow
+type Config interface {
+	/* Version */
+
+	// GetVersion returns the current project version
+	GetVersion() Version
+
+	/* String fields */
+
+	// GetDomain returns the project domain
+	GetDomain() string
+	// SetDomain sets the project domain
+	SetDomain(domain string) error
+
+	// GetRepository returns the project repository.
+	GetRepository() string
+	// SetRepository sets the project repository
+	SetRepository(repository string) error
+
+	// GetName returns the project name
+	// This method was introduced in project version 3-alpha.
+	GetName() string
+	// SetName sets the project name
+	// This method was introduced in project version 3-alpha.
+	SetName(name string) error
+
+	// GetLayout returns the config layout
+	// This method was introduced in project version 3-alpha.
+	GetLayout() string
+	// SetLayout sets the Config layout
+	// This method was introduced in project version 3-alpha.
+	SetLayout(layout string) error
+
+	/* Boolean fields */
+
+	// IsMultiGroup checks if multi-group is enabled
+	IsMultiGroup() bool
+	// SetMultiGroup enables multi-group
+	SetMultiGroup() error
+	// ClearMultiGroup disables multi-group
+	ClearMultiGroup() error
+
+	// IsComponentConfig checks if component config is enabled
+	// This method was introduced in project version 3-alpha.
+	IsComponentConfig() bool
+	// SetComponentConfig enables component config
+	// This method was introduced in project version 3-alpha.
+	SetComponentConfig() error
+	// ClearComponentConfig disables component config
+	// This method was introduced in project version 3-alpha.
+	ClearComponentConfig() error
+
+	/* Resources */
+
+	// ResourcesLength returns the number of tracked resources
+	ResourcesLength() int
+	// HasResource checks if the provided GVK is stored in the Config
+	HasResource(gvk resource.GVK) bool
+	// GetResource returns the stored resource matching the provided GVK
+	GetResource(gvk resource.GVK) (resource.Resource, error)
+	// AddResource adds the provided resource if it was not present, no-op if it was already present
+	AddResource(res resource.Resource) error
+	// UpdateResource adds the provided resource if it was not present, modifies it if it was already present
+	UpdateResource(res resource.Resource) error
+
+	// HasGroup checks if the provided group is the same as any of the tracked resources
+	HasGroup(group string) bool
+	// IsCRDVersionCompatible returns true if crdVersion can be added to the existing set of CRD versions.
```

Comment:
I am unable to get the plugins of a config. 
/hold

Reply:
I didn't expose that on purpose, what is the usage of this function? Plugins info is only stored if plugins decide to do so, so using this function will not really return if a plugins is being used, only if that plugin stored previously a configuration object.

Reply:
The motivation of the plugin implementation is to allow us to extend KB. 
I need to know all plugins of some config in many scenarios.  I can create tools based on the plugins used in the config, I can do customization and etc. 

In the above implementation, we need to check that to ensure that projects which were built before the plugins introduction will still working on as before.